File: impl/python/tests/test_carriage.py
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
"""C12 foreign carriage by class for the Python SDK, graded against the shared independent PER-CLASS
corpus. vectors/carriage/ is NOT a single cases.json: it is one subdirectory per carriage class --
vectors/carriage/{jsonrpc,http,msg,stream,doc,opaque}/cases.json -- and each carriage class is graded
against its OWN subdir's vectors. For every class the carriage body encodes to the oracle bytes and
recovers its foreign message byte-identical (octet-exact), proving the foreign field is carried
verbatim and never re-serialized, canonicalized, or rewritten (R-14.4/R-14.7).

N-AALP carries a foreign agent protocol by wrapping its message octet-for-octet in a signed N-AALP
carriage object; the OPAQUE class makes any protocol -- including an undefined one -- carriable on an
experimental protocol id with no registration (R-18.6). The carriage object's signer remains the
authority: a foreign principal named inside the foreign bytes never becomes an N-AALP authorization
identity (R-14.6), demonstrated in isolation over a real signed envelope object (NOT corpus-graded).

Written test-first; the carriage module is absent until ported, so this fails RED on import until
impl/python/naalp/carriage.py lands. Mutation: forcing the encoded class field (key 2) to the constant
0 in CarriageBody.to_value makes every non-JSONRPC class body diverge from the oracle and flips
test_per_class_octet_exact_round_trip.

Run:  python -m unittest tests.test_carriage      (from impl/python/)
"""
import csv
import json
import os
import unittest

from naalp import carriage, cbor, cose, envelope


def _repo_root():
    d = os.path.dirname(os.path.abspath(__file__))
    for _ in range(6):
        if os.path.isdir(os.path.join(d, "vectors", "carriage")):
            return d
        d = os.path.dirname(d)
    raise FileNotFoundError("vectors/carriage not found")


ROOT = _repo_root()


def _class_vectors(dir_):
    with open(os.path.join(ROOT, "vectors", "carriage", dir_, "cases.json"), encoding="utf-8") as f:
        return json.load(f)


# (subdir, expected class code) -- every carriage class present under vectors/carriage/.
CLASS_DIRS = [
    ("jsonrpc", carriage.CLASS_JSONRPC),
    ("http", carriage.CLASS_HTTP),
    ("msg", carriage.CLASS_MSG),
    ("stream", carriage.CLASS_STREAM),
    ("doc", carriage.CLASS_DOC),
    ("opaque", carriage.CLASS_OPAQUE),
]


def hb(s):
    return bytes.fromhex(s)


class CarriageConformance(unittest.TestCase):

    def test_per_class_octet_exact_round_trip(self):
        # MUTATION TARGET. R-14.7: each carriage class encodes to its own subdir's oracle bytes and
        # recovers its foreign message byte-identical (octet-exact).
        for dir_, want_class in CLASS_DIRS:
            c = _class_vectors(dir_)
            self.assertEqual(c["class"], want_class, "%s class code" % dir_)
            foreign = hb(c["foreign_hex"])
            cb = carriage.carry(c["protocol_id"], c["class"], c["content_type"],
                                hb(c["correlation_hex"]), c["method"], foreign)
            self.assertEqual(cb.bytes().hex(), c["body_hex"], "%s body bytes" % dir_)
            # Round-trip: decode the carriage body and recover the foreign octets exactly.
            rec = carriage.carriage_from_value(cbor.decode(cb.bytes()))
            self.assertEqual(rec.foreign, foreign, "%s foreign octet-exact" % dir_)
            self.assertEqual(rec.protocol_id, c["protocol_id"])
            self.assertEqual(rec.klass, c["class"])
            self.assertEqual(rec.method, c["method"])

    def test_opaque_undefined_protocol(self):
        # R-18.6: an undefined protocol carries under OPAQUE on an experimental protocol id (no
        # registration), byte-exact.
        c = _class_vectors("opaque")
        self.assertEqual(carriage.protocol_range(c["protocol_id"]), "experimental")
        blob = bytes([0x00, 0x01, 0x02, 0xFF, 0xFE, 0x7F, 0x80])
        cb = carriage.carry(c["protocol_id"], carriage.CLASS_OPAQUE, 1, b"", "", blob)
        rec = carriage.carriage_from_value(cbor.decode(cb.bytes()))
        self.assertEqual(rec.foreign, blob)

    def test_protocol_range_boundaries(self):
        # design §13.4 range classification, and the registry CSV's assigned ids are all standards.
        for id_, want in {0x00: "reserved", 0x01: "standards", 0x0F: "standards", 0x10: "experimental",
                          0x7F: "experimental", 0x80: "private", 0xFF: "private", 0x100: "invalid"}.items():
            self.assertEqual(carriage.protocol_range(id_), want, "range(%#x)" % id_)
        with open(os.path.join(ROOT, "vectors", "registry", "protocols.csv"), encoding="utf-8") as f:
            rows = list(csv.reader(f))
        self.assertGreater(len(rows), 1)
        for r in rows[1:]:  # header: protocol_id,name,class,range,reference
            id_ = int(r[0], 16)
            self.assertEqual(carriage.protocol_range(id_), "standards", "registry id %s" % r[0])
            self.assertEqual(r[3], "standards")

    def test_mapping_error_on_unknown_class(self):
        # R-14.8: an unrepresentable class is a typed mapping error, never a silent drop.
        with self.assertRaises(carriage.CarriageError) as cm:
            carriage.carry(0x10, 99, 0, b"", "x", b"y")
        self.assertEqual(cm.exception.kind, "MappingError")

    def test_not_delivered(self):
        # R-14.8: a below-foreign failure reports NotDelivered and never a false "delivered".
        with self.assertRaises(carriage.CarriageError) as cm:
            carriage.report(False)
        self.assertEqual(cm.exception.kind, "NotDelivered")
        rep = carriage.report(True)
        self.assertTrue(rep.delivered)

    def test_malformed_body_rejected(self):
        # A carriage body missing the mandatory foreign field (key 6) is rejected Malformed.
        m = cbor.M([(cbor.U(1), cbor.U(1)), (cbor.U(2), cbor.U(0)), (cbor.U(3), cbor.U(0)),
                    (cbor.U(4), cbor.B(b"")), (cbor.U(5), cbor.T("m"))])
        with self.assertRaises(carriage.CarriageError) as cm:
            carriage.carriage_from_value(m)
        self.assertEqual(cm.exception.kind, "Malformed")

    def test_identity_containment_isolation(self):
        # R-14.6 (NOT corpus-graded): a foreign principal named inside the foreign bytes confers no
        # authority; the authorizing principal is the N-AALP signer of the carriage object, over a real
        # signed envelope object. The foreign body recovers octet-exact from the signed object.
        seed = bytes([80]) * 32
        pk = cose.mldsa_keygen("ML-DSA-65", seed)
        foreign = b'{"jsonrpc":"2.0","method":"tools/call","params":{"from":"attacker-principal"}}'
        cb = carriage.carry(0x01, carriage.CLASS_JSONRPC, 0, bytes([1, 2, 3, 4]), "tools/call", foreign)
        obj = envelope.Object(kind=0, channel=13, signer=pk, created=100, effect=0,
                              body=cb.to_value(), tier=0, profile=cose.PROFILE_PUBLIC)
        signed = envelope.sign(obj, cose.ALG_MLDSA65, seed)
        o = envelope.verify(cose.PROFILE_PUBLIC, cose.ALG_MLDSA65, pk, lambda ch, k: True, signed)
        auth = carriage.carriage_authority(o)
        self.assertEqual(auth, pk)
        self.assertNotIn(b"attacker-principal", auth)
        rec = carriage.carriage_from_value(o.body)
        self.assertEqual(rec.foreign, foreign)
        self.assertIn(b"attacker-principal", rec.foreign)  # present but non-authoritative


if __name__ == "__main__":
    unittest.main()

File: impl/ruby/test/test_carriage.rb
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
#
# C12 foreign-carriage-by-class conformance for the Ruby SDK (design.md §13; R-14.1..14.8, R-18.6),
# graded against the shared independent PER-CLASS corpus vectors/carriage/<class>/cases.json (NOT one
# cases.json, and NOT produced by this code): each of the six carriage classes -- JSONRPC (0), HTTP (1),
# MSG (2), STREAM (3), DOC (4), OPAQUE (5) -- is graded against its OWN subdirectory oracle
# (vectors/carriage/{jsonrpc,http,msg,stream,doc,opaque}/cases.json). The tests grade the deterministic
# carriage-body bytes byte-for-byte for every class, the octet-for-octet recovery of the foreign
# message (R-14.4), the protocol-id ranges, the fail-closed class/foreign rejections, the honest
# delivery report, and the signer-is-authority rule.
#
# The LOAD-BEARING C12 property graded here: the foreign message is carried VERBATIM and MUST NOT be
# re-serialized, canonicalized, summarized, or rewritten (R-14.4); N-AALP metadata is carried AROUND it.
#
# Written test-first; the module is absent until ported, so this fails RED (uninitialized constant
# Naalp::Carriage) until impl/ruby/lib/naalp/carriage.rb lands, and a mutation rewriting the carried
# foreign field to empty flips test_carriage_bodies_match_oracle on its assertion.
#
# Run:  ruby -Ilib -Itest test/test_carriage.rb      (from impl/ruby/)
require 'minitest/autorun'
require 'json'
require 'naalp'

# subdir name -> expected numeric class code (the per-class oracle layout).
CARRIAGE_CLASSES = { "jsonrpc" => 0, "http" => 1, "msg" => 2, "stream" => 3, "doc" => 4, "opaque" => 5 }.freeze

def carriage_class_vectors(klass)
  d = File.expand_path(File.dirname(__FILE__))
  6.times do
    p = File.join(d, "vectors", "carriage", klass, "cases.json")
    return JSON.parse(File.read(p, encoding: "utf-8")) if File.file?(p)
    d = File.dirname(d)
  end
  raise "vectors/carriage/#{klass}/cases.json not found"
end

def hb(hex)
  [hex].pack("H*")
end

class CarriageConformance < Minitest::Test
  CLASSES = CARRIAGE_CLASSES.keys.each_with_object({}) { |k, h| h[k] = carriage_class_vectors(k) }

  def body_from(cv)
    Naalp::Carriage::CarriageBody.new(
      cv["protocol_id"], cv["class"], cv["content_type"],
      hb(cv["correlation_hex"]), cv["method"], hb(cv["foreign_hex"])
    )
  end

  def test_class_vocabulary
    assert_equal 0, Naalp::Carriage::CLASS_JSONRPC
    assert_equal 1, Naalp::Carriage::CLASS_HTTP
    assert_equal 2, Naalp::Carriage::CLASS_MSG
    assert_equal 3, Naalp::Carriage::CLASS_STREAM
    assert_equal 4, Naalp::Carriage::CLASS_DOC
    assert_equal 5, Naalp::Carriage::CLASS_OPAQUE
    %w[JSONRPC HTTP MSG STREAM DOC OPAQUE].each_with_index { |name, code| assert_equal name, Naalp::Carriage.class_name(code) }
    assert_equal "unknown", Naalp::Carriage.class_name(6)
  end

  # THE MUTATION TARGET. Ruby encoding == the non-circular oracle, byte-for-byte, for ALL SIX carriage
  # classes, EACH graded against its OWN per-class subdirectory oracle. The corpus `class` field must
  # equal the subdir's expected class code. Rewriting the carried foreign field to empty flips this on
  # its per-class assertion.
  def test_carriage_bodies_match_oracle
    CLASSES.each do |klass, cv|
      assert_equal CARRIAGE_CLASSES[klass], cv["class"], "#{klass} corpus class code"
      body = body_from(cv)
      assert_equal cv["body_hex"], body.bytes.unpack1("H*"), "#{klass} carriage body (vectors/carriage/#{klass})"
    end
  end

  # The load-bearing R-14.4 property: a carriage body decodes back to the foreign message OCTET-FOR-
  # OCTET, and re-encodes to the identical bytes -- N-AALP never re-serializes or rewrites the foreign
  # payload. Graded for every class against its own subdir oracle.
  def test_foreign_carried_octet_for_octet
    CLASSES.each do |klass, cv|
      parsed = Naalp::Carriage.carriage_from_value(Naalp::CBOR.decode(hb(cv["body_hex"])))
      assert_equal cv["foreign_hex"], parsed.foreign.unpack1("H*"), "#{klass} foreign recovered octet-for-octet"
      assert_equal cv["method"], parsed.method, "#{klass} method recovered"
      assert_equal cv["class"], parsed.klass, "#{klass} class recovered"
      assert_equal cv["protocol_id"], parsed.protocol_id, "#{klass} protocol_id recovered"
      # round-trip re-encode is byte-identical (verbatim carriage, R-14.4).
      assert_equal cv["body_hex"], parsed.bytes.unpack1("H*"), "#{klass} round-trip re-encode is byte-identical"
    end
  end

  def test_validate_class_and_protocol_range
    # a class code outside 0..5 is a MappingError, never a silent drop (R-14.8).
    err = assert_raises(Naalp::Carriage::CarriageError) { Naalp::Carriage.validate_class(6) }
    assert_equal "MappingError", err.kind
    assert_nil Naalp::Carriage.validate_class(Naalp::Carriage::CLASS_OPAQUE)
    # protocol-id ranges (design §13.4): the corpus uses standards (<=0x0F) and experimental (0x10..).
    assert_equal "reserved", Naalp::Carriage.protocol_range(0x00)
    assert_equal "standards", Naalp::Carriage.protocol_range(CLASSES["doc"]["protocol_id"])       # 2
    assert_equal "experimental", Naalp::Carriage.protocol_range(CLASSES["msg"]["protocol_id"])     # 16 = 0x10
    assert_equal "experimental", Naalp::Carriage.protocol_range(CLASSES["opaque"]["protocol_id"])  # 17 = 0x11
    assert_equal "private", Naalp::Carriage.protocol_range(0x80)
    assert_equal "invalid", Naalp::Carriage.protocol_range(0x100)
  end

  def test_carriage_from_value_rejects_malformed
    # a non-map value is Malformed.
    err = assert_raises(Naalp::Carriage::CarriageError) { Naalp::Carriage.carriage_from_value(Naalp::CBOR::U.new(1)) }
    assert_equal "Malformed", err.kind
    # a body missing the mandatory foreign field (key 6) is Malformed.
    no_foreign = Naalp::CBOR::M.new([
      [Naalp::CBOR::U.new(1), Naalp::CBOR::U.new(1)],
      [Naalp::CBOR::U.new(2), Naalp::CBOR::U.new(0)],
      [Naalp::CBOR::U.new(3), Naalp::CBOR::U.new(0)],
      [Naalp::CBOR::U.new(4), Naalp::CBOR::B.new("".b)],
      [Naalp::CBOR::U.new(5), Naalp::CBOR::T.new("m")],
    ])
    err2 = assert_raises(Naalp::Carriage::CarriageError) { Naalp::Carriage.carriage_from_value(no_foreign) }
    assert_equal "Malformed", err2.kind
    # a well-formed body carrying an out-of-range class (6) is a MappingError.
    bad_class = Naalp::CBOR::M.new([
      [Naalp::CBOR::U.new(1), Naalp::CBOR::U.new(1)],
      [Naalp::CBOR::U.new(2), Naalp::CBOR::U.new(6)],
      [Naalp::CBOR::U.new(3), Naalp::CBOR::U.new(0)],
      [Naalp::CBOR::U.new(4), Naalp::CBOR::B.new("".b)],
      [Naalp::CBOR::U.new(5), Naalp::CBOR::T.new("m")],
      [Naalp::CBOR::U.new(6), Naalp::CBOR::B.new("x".b)],
    ])
    err3 = assert_raises(Naalp::Carriage::CarriageError) { Naalp::Carriage.carriage_from_value(bad_class) }
    assert_equal "MappingError", err3.kind
  end

  def test_report_never_false_delivered
    rep = Naalp::Carriage.report(true)
    assert rep.delivered, "a real delivery reports delivered"
    err = assert_raises(Naalp::Carriage::CarriageError) { Naalp::Carriage.report(false) }
    assert_equal "NotDelivered", err.kind
  end

  # The authorizing principal of a carriage object is the N-AALP SIGNER of the object (envelope field
  # 5), never any foreign principal named inside the foreign bytes (R-14.6). Demonstrated on a real
  # Naalp::Envelope::Object.
  def test_carriage_authority_is_the_naalp_signer
    cv = CLASSES["jsonrpc"]
    body = body_from(cv)
    signer = ("\xAB" * 32).b
    obj = Naalp.object(kind: 1, channel: Naalp::Carriage::CLASS_JSONRPC, signer: signer,
                       created: 1785000000000, effect: Naalp::Policy::READ_ONLY,
                       body: Naalp::CBOR.decode(body.bytes))
    assert_equal signer, Naalp::Carriage.carriage_authority(obj), "authority is the N-AALP signer, not any foreign principal"
  end
end
